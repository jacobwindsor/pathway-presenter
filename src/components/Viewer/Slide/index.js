import React from 'react';
import { Pvjs } from '@wikipathways/pvjs';
import { Slide } from 'spectacle';
import './index.css';

export default (props) => {
    const manipulator = props.manipulator;
    const slide = props.slide;
    console.log(slide);
    slide.targets.forEach(singleTarget => {
        if (singleTarget.highlighted) manipulator.highlightOn(singleTarget.entityId, singleTarget.highlightedColor)
    });

    return (
        <Slide key={`slide-${props.slide.id}`} />
    )
}