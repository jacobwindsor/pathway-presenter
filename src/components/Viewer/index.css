.presentation-viewer {
    width: 100%;
    height: 100%;
    overflow: hidden;
    position: relative;
}

.presentation-viewer .controls {
    position: absolute;
    bottom: 0;
    right: 0;
}