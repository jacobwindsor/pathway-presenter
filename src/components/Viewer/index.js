import React, { Component } from 'react';
import presentations from '../../data/presentations';
import './index.css';
import Diagram from '../Diagram';
import Controls from './components/Controls';
import Loading from '../Loading';
import ErrorMessage from '../ErrorMessage';
import PropTypes from 'prop-types';
import 'roboto-fontface/css/roboto/roboto-fontface.css';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import Title from '../Title';
import TitleSlide from './components/TitleSlide';
import * as screenfull from 'screenfull';
import { cloneDeep } from 'lodash';

class Viewer extends Component {
    constructor(props) {
        super(props);
        this.state = {
            loading: true,
            isFullScreen: false,
        };
    }

    componentDidMount() {
        const { presId } = this.props;
        presentations.get(presId)
            .then(presentation => {
                this.addEventListeners();

                const copy = cloneDeep(presentation);
                if (copy.title) {
                    copy.slides = [{
                        isTitleSlide: true,
                        title: copy.title,
                        authorName: copy.authorName,
                    }].concat(copy.slides);
                }
                this.setState({
                    presentation: copy,
                    loading: false,
                    activeSlideIndex: 0
                });
            })
            .catch(err => {
                this.setState({
                    error: {
                        message: err.message
                    },
                    loading: false
                })
            });
    }

    addEventListeners() {
        window.addEventListener('keydown', (e) => {
            switch(e.keyCode) {
                case 39:
                    // right arrow
                    this.nextSlide();
                    break;
                case 37:
                    // Left arrow
                    this.prevSlide()
            }
        })
    }

    nextSlide = () =>  {
        const { activeSlideIndex, presentation } = this.state;
        if (activeSlideIndex === presentation.slides.length - 1) return;
        this.setState({
            activeSlideIndex: activeSlideIndex + 1,
        })
    };

    prevSlide = () => {
        const { activeSlideIndex } = this.state;
        if (activeSlideIndex === 0) return;
        this.setState({
            activeSlideIndex: activeSlideIndex - 1,
        })
    };

    onReady = () => {
        this.setState({loading: false})
    };

    handleToggleFullScreen = () => {
        if(screenfull.enabled) {
            screenfull.toggle();
            screenfull.onchange(() => {
                this.setState({
                    isFullScreen: screenfull.isFullscreen,
                })
            })
        }
    };

    render() {
        const { loading, presentation, activeSlideIndex, error, isFullScreen } = this.state;

        if (error)  {
            return <ErrorMessage message={error.message} />
        }

        if ( loading ) return  <Loading />;
        if (! presentation) return null;

        const activeSlide = presentation.slides[activeSlideIndex];
        const diagramSlide = activeSlide.isTitleSlide ? presentation.slides[activeSlideIndex + 1] : activeSlide;

        const isHidden = activeSlide.isTitleSlide;

        return (
            <MuiThemeProvider>
                <div className="presentation-viewer">
                    { activeSlide.isTitleSlide ? <TitleSlide
                        title={activeSlide.title}
                        authorName={activeSlide.authorName} />
                        : null }

                    { !activeSlide.isTitleSlide && activeSlide.title ? <Title title={activeSlide.title} /> : null }

                    <Diagram
                        wpId={presentation.wpId}
                        version={presentation.version}
                        slide={diagramSlide}
                        showPanZoomControls={false}
                        isHidden={isHidden}
                        onReady={this.onReady}  />

                    <Controls
                        onBackward={this.prevSlide}
                        onForward={this.nextSlide}
                        handleToggleFullScreen={this.handleToggleFullScreen}
                        isFullScreen={isFullScreen}
                    />
                </div>
            </MuiThemeProvider>
        );
    }
}

Viewer.propTypes = {
    presId: PropTypes.string.isRequired
};

export default Viewer;