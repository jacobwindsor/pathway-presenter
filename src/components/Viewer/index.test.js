import React from 'react';
import { shallow } from 'enzyme';
import Viewer from './index';
import toJson from 'enzyme-to-json';

describe('The Viewer. This shows presentations in "presentation mode"', () => {
  it('renders without crashing', () => {
    const comp = shallow(<Viewer presId="1234" />).find('Viewer').shallow();
    expect(toJson(comp)).toMatchSnapshot();
  });

  it('should show a loading spinner', () => {
    const comp = shallow(<Viewer presId="1234" />).find('Viewer').shallow();
    expect(comp.dive().name()).toEqual('Spinner');
    expect(comp.dive().is('.loading-spinner')).toBeTruthy();
  });

  it('should show a presentation', () => {
    const comp = shallow(<Viewer presId="1234" />).find('Viewer').shallow();
    comp.setState({
      presentation: {
        wpId: 'WP4',
        version: 0,
        title: 'My pres',
        author: 'Nanny McPhee',
        slides: [
          {
            title: 'First slide',
            targets: []
          }
        ]
      },
      loading: false,
      error: null,
      activeSlideIndex: 0
    });

    expect(comp.find('.loading-spinner')).toHaveLength(0);
    expect(comp.find('.presentation-viewer')).toHaveLength(1);
    expect(comp.find('Controls')).toHaveLength(1);

    expect(toJson(comp)).toMatchSnapshot();
  });
});
