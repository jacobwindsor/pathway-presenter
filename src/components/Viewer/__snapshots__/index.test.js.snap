// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Viewer. This is shows presentations in "presentation mode" renders without crashing 1`] = `<Component />`;

exports[`The Viewer. This is shows presentations in "presentation mode" should show a presentation 1`] = `
<MuiThemeProvider>
  <div
    className="presentation-viewer"
  >
    <Component
      title="First slide"
    />
    <Diagram
      detailPanelEnabled={false}
      hidden={false}
      onReady={[Function]}
      showPanZoomControls={false}
      slide={
        Object {
          "targets": Array [],
          "title": "First slide",
        }
      }
      version={0}
      wpId="WP4"
    />
    <Controls
      handleToggleFullScreen={[Function]}
      isFullScreen={false}
      onBackward={[Function]}
      onForward={[Function]}
    />
  </div>
</MuiThemeProvider>
`;
